improver/generate_ancillaries/generate_timezone_mask.py

```diff
@@ -0,0 +1,342 @@
+        only relevant if using daylights savings. The attribute
+        includes_daylights_savings is set to indicate this.
+
+        Args:
+            cube (iris.cube.Cube):
+                A cube with the desired grid from which coordinates are taken
+                for inclusion in the template.
+        Returns:
+            iris.cube.Cube:
+                A template cube in which each timezone mask can be stored.
+        """
+        time_point = np.array(self.time.timestamp(), dtype=np.int64)
+        time_coord = iris.coords.DimCoord(
+            time_point,
+            "time",
+            units=Unit("seconds since 1970-01-01 00:00:00", calendar="gregorian"),
+        )
+
+        for crd in cube.coords(dim_coords=False):
+            cube.remove_coord(crd)
+        cube.add_aux_coord(time_coord)
+
+        attributes = generate_mandatory_attributes([cube])
+        attributes["includes_daylights_savings"] = str(not self.ignore_dst)
+
+        return create_new_diagnostic_cube(
+            "timezone_mask", 1, cube, attributes, dtype=np.int32
+        )
+
+    def _group_timezones(self, timezone_mask):
+        """
+        If the ancillary will be used with data that is not available at hourly
+        intervals, the masks can be grouped to match the intervals of the data.
+        For example, 3-hourly interval data might group UTC offsets:
+
+            {12: [-12, -11], 9: [-10, -9, -8], 6: [-7, -6, -5], etc.}
+
+        The dictionary specifying the groupings has a key, which provides the
+        UTC offset to be used for the group. The list contains the UTC offsets
+        that should be grouped together.
+
+        The grouped UTC_offset cubes are collapsed together over the UTC_offset
+        coordinate using iris.analysis.MIN. This means all the unmasked (0)
+        points in each cube are preserved as the dimension is collapsed,
+        enlarging the unmasked region to include all unmaksed points from all
+        the cubes.
+
+        Args:
+            timezone_mask (iris.cube.CubeList):
+                A cube list containing a mask cube for each UTC offset that
+                has been found necessary.
+        Returns:
+            iris.cube.CubeList:
+                A cube list containing cubes created by blending together
+                different UTC offset cubes to create larger masked regions.
+        """
+        grouped_timezone_masks = iris.cube.CubeList()
+        for offset, group in self.groupings.items():
+            constraint = iris.Constraint(
+                UTC_offset=lambda cell: group[0] <= cell <= group[-1]
+            )
+            subset = timezone_mask.extract(constraint)
+            if not subset:
+                continue
+            subset = subset.merge_cube()
+            if subset.coord("UTC_offset").shape[0] > 1:
+                subset = collapsed(subset, "UTC_offset", iris.analysis.MIN)
+                subset.coord("UTC_offset").points = [offset]
```

Comment:
As the keys within the `groupings` dictionary could be any number, does there need to be a check that the key is within the range given by the points?

Reply:
Done, with a test added.